utxoscanner.go

```diff
@@ -0,0 +1,517 @@
+func (s *UtxoScanner) Enqueue(req *GetUtxoRequest) {
+	log.Debugf("Enqueuing request for %s with start height %d",
+		req.OutPoint.String(), req.StartHeight)
+	s.cv.L.Lock()
+
+	// Insert the request into the queue and signal any threads that might be
+	// waiting for new elements.
+	heap.Push(&s.pq, req)
+
+	s.cv.L.Unlock()
+	s.cv.Signal()
+}
+
+// NOTE: This method MUST be spawned as a goroutine.
+func (s *UtxoScanner) batchManager() {
+	defer s.wg.Done()
+
+	for {
+		s.cv.L.Lock()
+		if s.stopQueue {
+			s.cv.L.Unlock()
+			return
+		}
+
+		// Re-queue previously skipped requests for next batch.
+		for _, request := range s.nextBatch {
+			heap.Push(&s.pq, request)
+		}
+
+		s.nextBatch = nil
+
+		s.cv.L.Unlock()
+
+		failedRequests, err := s.scan()
+
+		// If there was an error, then notify the currently outstanding
+		// requests.
+		if err != nil {
+			for _, request := range failedRequests {
+				request.Result(nil, err)
+			}
+		}
+	}
+}
+
+// Returns the GetUtxoRequest with the lowest block height. If no elements are
+// available, then block until one is added.
+func (s *UtxoScanner) peek() *GetUtxoRequest {
+	s.cv.L.Lock()
+	defer s.cv.L.Unlock()
+
+	// Block until the queue is no longer empty.
+	for !s.stopQueue && s.pq.IsEmpty() {
+		s.cv.Wait()
+	}
+
+	// We return nil only in the case that we've been interrupted, so callers
+	// can use this to determine that the UtxoScanner is shutting down.
+	if s.stopQueue {
+		return nil
+	}
+
+	return s.pq.Peek()
+}
+
+// dequeueAtHeight returns all GetUtxo requests that have starting height of the
+// given height.
+func (s *UtxoScanner) dequeueAtHeight(height uint32) []*GetUtxoRequest {
+	s.cv.L.Lock()
+	defer s.cv.L.Unlock()
+
+	// Take any requests that are too old to go in this batch and keep them for
+	// the next batch.
+	for !s.pq.IsEmpty() && s.pq.Peek().StartHeight < height {
+		item := heap.Pop(&s.pq).(*GetUtxoRequest)
+		s.nextBatch = append(s.nextBatch, item)
+	}
+
+	var requests []*GetUtxoRequest
+	for !s.pq.IsEmpty() && s.pq.Peek().StartHeight == height {
+		item := heap.Pop(&s.pq).(*GetUtxoRequest)
+		requests = append(requests, item)
+	}
+
+	return requests
+}
+
+// findInitialTransactions returns a SpendReport for the UTXO, or nil if it does
+// not exist in this block.
+func findInitialTransactions(block *wire.MsgBlock,
+	newReqs []*GetUtxoRequest) map[wire.OutPoint]*SpendReport {
+
+	// Construct a reverse index from txid to all a list of requests whose
+	// outputs share the same txid.
+	txidReverseIndex := make(map[chainhash.Hash][]*GetUtxoRequest)
+	for _, req := range newReqs {
+		txidReverseIndex[req.OutPoint.Hash] = append(
+			txidReverseIndex[req.OutPoint.Hash], req,
+		)
+	}
+
+	// Iterate over the transactions in this block, hashing each and
+	// querying our reverse index to see if any requests depend on the txn.
+	initialTxns := make(map[wire.OutPoint]*SpendReport)
+	for _, tx := range block.Transactions {
+		txidReqs, ok := txidReverseIndex[tx.TxHash()]
```

Comment:
What's the rationale behind this logic? The previous outpoint will appear in the _input_ of each of the transactions. Therefore, it shouldn't be the case that a transaction matches the reverse filter unless it's the transaction that _created_ the output.

Reply:
Yeh, this method searches for the transactions that create the original output on chain. The method description could use some more specifics, will update.